File: frontend/src/lib/sheets/calculations/roomBill.calc.ts
import type { OtherFee, Room, RoomBill } from '../../../types/api';

export interface RoomBillInput {
  roomId: string;
  billingMonth: string;
  electricityOldReading: number;
  electricityNewReading: number;
  waterOldReading: number;
  waterNewReading: number;
  wifiFee?: number;
  trashFee?: number;
  monthlyRent?: number;
  otherFees?: OtherFee[];
  note?: string;
}

export const buildRoomBillPayload = (
  input: RoomBillInput,
  room: Room,
): Omit<RoomBill, '_id' | 'createdAt' | 'updatedAt'> & { roomId: string } => {
  const electricityUsed = Math.max(
    0,
    input.electricityNewReading - input.electricityOldReading,
  );
  const waterUsed = Math.max(0, input.waterNewReading - input.waterOldReading);
  const electricityAmount = electricityUsed * room.electricityUnitPrice;
  const waterAmount = waterUsed * room.waterUnitPrice;
  const wifiFee = input.wifiFee ?? room.wifiFee;
  const trashFee = input.trashFee ?? room.trashFee;
  const monthlyRent = input.monthlyRent ?? room.monthlyRent;
  const otherFees = input.otherFees ?? [];
  const otherFeesTotal = otherFees.reduce((sum, fee) => sum + (fee.amount ?? 0), 0);

  const totalAmount =
    electricityAmount + waterAmount + wifiFee + trashFee + monthlyRent + otherFeesTotal;

  return {
    roomId: room._id,
    billingMonth: input.billingMonth,
    electricityOldReading: input.electricityOldReading,
    electricityNewReading: input.electricityNewReading,
    electricityUsed,
    waterOldReading: input.waterOldReading,
    waterNewReading: input.waterNewReading,
    waterUsed,
    electricityAmount,
    waterAmount,
    wifiFee,
    trashFee,
    monthlyRent,
    otherFees,
    note: input.note ?? '',
    totalAmount,
  };
};

File: mobile/lib/sheets/calculations/room_bill_calc.dart
import 'dart:math' as math;

import '../../models/api_models.dart';

class RoomBillInput {
  const RoomBillInput({
    required this.roomId,
    required this.billingMonth,
    required this.electricityOldReading,
    required this.electricityNewReading,
    required this.waterOldReading,
    required this.waterNewReading,
    this.wifiFee,
    this.trashFee,
    this.monthlyRent,
    this.otherFees,
    this.note,
  });

  final String roomId;
  final String billingMonth;
  final double electricityOldReading;
  final double electricityNewReading;
  final double waterOldReading;
  final double waterNewReading;
  final double? wifiFee;
  final double? trashFee;
  final double? monthlyRent;
  final List<OtherFee>? otherFees;
  final String? note;
}

typedef RoomBillComputed = ({
  String roomId,
  String billingMonth,
  double electricityOldReading,
  double electricityNewReading,
  double electricityUsed,
  double waterOldReading,
  double waterNewReading,
  double waterUsed,
  double electricityAmount,
  double waterAmount,
  double wifiFee,
  double trashFee,
  double monthlyRent,
  List<OtherFee> otherFees,
  String note,
  double totalAmount,
});

RoomBillComputed buildRoomBillPayload(RoomBillInput input, Room room) {
  final electricityUsed = math
      .max(0, input.electricityNewReading - input.electricityOldReading)
      .toDouble();
  final waterUsed =
      math.max(0, input.waterNewReading - input.waterOldReading).toDouble();
  final electricityAmount = electricityUsed * room.electricityUnitPrice;
  final waterAmount = waterUsed * room.waterUnitPrice;
  final wifiFee = input.wifiFee ?? room.wifiFee;
  final trashFee = input.trashFee ?? room.trashFee;
  final monthlyRent = input.monthlyRent ?? room.monthlyRent;
  final otherFees = input.otherFees ?? <OtherFee>[];
  final otherFeesTotal =
      otherFees.fold<double>(0, (sum, fee) => sum + fee.amount);

  final totalAmount = electricityAmount +
      waterAmount +
      wifiFee +
      trashFee +
      monthlyRent +
      otherFeesTotal;

  return (
    roomId: room.id,
    billingMonth: input.billingMonth,
    electricityOldReading: input.electricityOldReading,
    electricityNewReading: input.electricityNewReading,
    electricityUsed: electricityUsed,
    waterOldReading: input.waterOldReading,
    waterNewReading: input.waterNewReading,
    waterUsed: waterUsed,
    electricityAmount: electricityAmount,
    waterAmount: waterAmount,
    wifiFee: wifiFee,
    trashFee: trashFee,
    monthlyRent: monthlyRent,
    otherFees: otherFees,
    note: input.note ?? '',
    totalAmount: totalAmount,
  );
}
